kube-runtime/src/controller/mod.rs

```diff
@@ -258,56 +265,121 @@ where
+                                    request.obj_ref.clone(),
+                                    scheduler_tx,
+                                )
+                                // Reconciler errors are OK from the applier's PoV, we need to apply the error policy
+                                // to them separately
+                                .map(|res| Ok((request.obj_ref, res)))
+                            })
+                            .instrument(reconciler_span)
+                            .left_future()
+                    }
+                    None => future::err(Error::ObjectNotFound(request.obj_ref.erase())).right_future(),
                 }
             })
             .on_complete(async { tracing::debug!("applier runner terminated") })
         },
     )
     .on_complete(async { tracing::debug!("applier runner-merge terminated") })
     // finally, for each completed reconcile call:
-    .and_then(move |(obj_ref, reconciler_result, reconciler_span)| {
-        let (Action { requeue_after }, requeue_reason) = match &reconciler_result {
-            Ok(action) =>
-                // do what user told us
-                (action.clone(), ReconcileReason::ReconcilerRequestedRetry),
-            Err(err) =>
-                // reconciler fn call failed
-                (reconciler_span.in_scope(|| error_policy(err, err_context.clone())), ReconcileReason::ErrorPolicyRequestedRetry),
-        };
-        let mut scheduler_tx = scheduler_tx.clone();
-        async move {
-            // Transmit the requeue request to the scheduler (picked up again at top)
-            if let Some(delay) = requeue_after {
-                // Failure to schedule item = in graceful shutdown mode, ignore
-                let _ = scheduler_tx
-                    .send(ScheduleRequest {
-                        message: ReconcileRequest {obj_ref: obj_ref.clone(), reason: requeue_reason},
-                        run_at: Instant::now() + delay,
-                    })
-                    .await;
-            }
-            match reconciler_result {
-                Ok(action) => Ok((obj_ref, action)),
-                Err(err) => Err(Error::ReconcilerFailed(err, obj_ref.erase()))
-            }
+    .and_then(move |(obj_ref, reconciler_result)| async move {
+        match reconciler_result {
+            Ok(action) => Ok((obj_ref, action)),
+            Err(err) => Err(Error::ReconcilerFailed(err, obj_ref.erase())),
         }
     })
     .on_complete(async { tracing::debug!("applier terminated") })
 }
 
+/// Internal helper that runs post-reconciliation (such as requesting rescheduling) tasks in the scheduled context of the reconciler
+///
+/// This could be an `async fn`, but isn't because we want it to be [`Unpin`]
+#[pin_project]
+#[must_use]
+struct PostReconciler<K: Resource, ReconcilerErr> {
+    reschedule_tx: channel::mpsc::Sender<ScheduleRequest<ReconcileRequest<K>>>,
```

Comment:
this struct does seem to add a lot of code here for the fairly simple Future::poll body it implements. since you are going to all this trouble to write a much more complex setup with a manual future, what problem does this avoid?

Reply:
hm, it does look like you have managed to get rid of some code in the final `and_then`. some of it got lifted above the `on_complete` trace line, and some of it into the new `PostReconciler`. i need to look more at this tomorrow, lots of stuff moving around here :(

as a side-note: if you are able to document something like what this struct is doing / what it is responsible for / what it is encapsulating, that might make the complexity a little more easier to stomach.

Reply:
`async` blocks (including the ones implied by `async fn`) are always `!Unpin`. `Executor` requires all of its tasks to be `Unpin` because of how it stores its tasks internally.

To combine the two we'd need to `Box::pin` (or worse, `tokio::spawn`) it, costing an extra allocation.

Reply:
Ok this makes sense. I read this now in a normal full page view source and it's a lot cleaner. Github's split diff is not great for these lines.

It feels like can reasonable be renamed to `Rescheduler`, even if it does an `error_policy` callout (because this callout is to how to reschedule)